cloud/ec2_client.go

```diff
@@ -921,6 +926,65 @@ func (c *awsClientImpl) GetPublicDNSName(ctx context.Context, h *host.Host) (str
 	return *instance.PublicDnsName, nil
 }
 
+func (c *awsClientImpl) AllocateAddress(ctx context.Context, input *ec2.AllocateAddressInput) (*ec2.AllocateAddressOutput, error) {
+	retryOpts := awsClientDefaultRetryOptions()
+	// Use fewer attempts to allocate an address because this is just an
+	// optimization to attempt to reduce costs for using public IPv4 addresses
+	// for hosts.
+	retryOpts.MaxAttempts = 3
+	var output *ec2.AllocateAddressOutput
+	var err error
+	err = utility.Retry(
+		ctx,
+		func() (bool, error) {
+			msg := makeAWSLogMessage("AllocateAddress", fmt.Sprintf("%T", c), input)
+			output, err = c.ec2Client.AllocateAddress(ctx, input)
+			if err != nil {
+				var apiErr smithy.APIError
+				if errors.As(err, &apiErr) {
+					grip.Debug(message.WrapError(apiErr, msg))
+				}
+				if strings.Contains(apiErr.Error(), ec2InsufficientAddressCapacity) || strings.Contains(apiErr.Error(), ec2AddressLimitExceeded) {
+					return false, err
+				}
+				return true, err
+			}
+			grip.Info(msg)
+			return false, nil
+		}, retryOpts)
+	if err != nil {
+		return nil, err
+	}
+	return output, nil
+}
+
+func (c *awsClientImpl) AssociateAddress(ctx context.Context, input *ec2.AssociateAddressInput) (*ec2.AssociateAddressOutput, error) {
+	var output *ec2.AssociateAddressOutput
+	var err error
+	err = utility.Retry(
+		ctx,
+		func() (bool, error) {
+			msg := makeAWSLogMessage("AssociateAddress", fmt.Sprintf("%T", c), input)
+			output, err = c.ec2Client.AssociateAddress(ctx, input)
+			if err != nil {
+				var apiErr smithy.APIError
+				if errors.As(err, &apiErr) {
+					grip.Debug(message.WrapError(apiErr, msg))
+				}
+				if strings.Contains(err.Error(), ec2ResourceAlreadyAssociated) {
```

Comment:
is it possible that this could return that it's already associated to the resource we're asking about ,in which case we wouldn't want to error?

Reply:
This is only returned if it's currently associated with a different instance. The [docs](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ec2@v1.180.0#Client.AssociateAddress) say this is idempotent, so if you try associating it with an instance it's already associated with, it'll return success.